github/internals/entities.py

```diff
@@ -0,0 +1,768 @@
+            return self.quit()
+
+        self.done = True
+        self.aborted = True
+
+    def quit(self):
+        sys.exit()
+
+    def register_task(self, task):
+        self.task = task
+
+    def unregister_task(self):
+        self.task = None
+
+
+class JobResult(Stateful):
+    def __init__(
+        self, state: State, description: Text="", url: Text=""
+    ) -> None:
+        if state not in self.valid_states:
+            raise ValueError('invalid state: {}'.format(state))
+
+        self.state = state
+        self.description = description[:GITHUB_DESCRIPTION_LIMIT]
+        self.url = url
+
+
+class JobDispatcher(AbcCallable):
+    def __init__(
+        self, job_data: Dict, build_target: Tuple[Text, Text]
+    ) -> None:
+        """Constructs a job for a runner from a given job dictionary
+
+        The job dictionary should have the following structure:
+
+        class: RunPytest
+        args:
+            build_url: '{fedora-27/build_url}'
+            test_suite: path/to/test.py
+            template: *ci - master - f27
+            timeout: 3600
+            topology:
+                name: topology_name
+                cpu: 4
+                memory: 5750
+        """
+        self.task_class = getattr(
+            __import__('tasks', fromlist=[job_data['class']]),
+            job_data['class']
+        )
+        self.kwargs = job_data['args']
+        self.kwarg_lookup = {
+            'git_repo': build_target[0],
+            'git_refspec': build_target[1]
+        }
+
+    @property
+    def timeout(self) -> int:
+        return self.kwargs.get('timeout') or 0
+
+    def __call__(self, dependencies_results: Dict=None) -> JobResult:
+        """Calls the constructed job and waits for its result"""
+        state = State.FAILURE
+        description = ""
+
+        # As we can have dependencies, obviously, we will need theirs results
+        # For example, URL with RPM packages
+        if dependencies_results is not None:
+            for task_name, result in dependencies_results.items():
+                self.kwarg_lookup[
+                    "{}_description".format(task_name)
+                ] = result.description
+                self.kwarg_lookup["{}_url".format(task_name)] = result.url
+
+        kwargs = {}
+        for key, value in self.kwargs.items():
+            if isinstance(value, str):
+                value = value.format(**self.kwarg_lookup)
+            kwargs[key] = value
+
+        job = self.task_class(**kwargs)
+        try:
+            job()
+        except TaskException as e:
+            description = str(e)
+            state = State.ERROR
+            sentry_report_exception({"module": "tasks"})
+        else:
+            description = job.description
+            if job.returncode == 0:
+                state = State.SUCCESS
+            else:
+                state = State.FAILURE
+        finally:
+            return JobResult(state, description, job.remote_url)
```

Comment:
I don't think this particular return should be in `finally` block as it would eat any exception raised in `except`/`else` blocks (although there aren't many points where that could happen right now). This was quite interesting for me to learn, actually.
Try running the following snippet in the python interpreter if you want:
```python
import sys

def x():
        try:
                raise RuntimeError('something')
        except Exception as e:
                raise RuntimeError('something2')
        finally:
                return 5

print(x())
print(sys.exc_info())
```
The effort to really run the `return` in the `finally` block is so high that the raised exception is eaten.